### DISCUSSION TITLE
DX for $fetch

### DISCUSSION THREAD

**Discussion by cawa-93** [nuxt/nuxt#17560] [Ideas]
Fetch looks fabulous. However, there are things that I lack in the area of error handling

Request a non-existent route. This does not indicate a TS error. However, it returns type `unknown`, which reduces the likelihood of error
```ts
$fetch('/api/404')
```

No validation of the request body. It would be extremely useful to have hints such as the required fields of the query body.
```ts
$fetch('/api/update', {body: {
id: 'not-a-number' // Should show TS Error
}})
```

Error type checking now looks a bit cumbersome. I lack consistency in imports. I want auto import to work, or import everything you need from #imports. I'm missing a handy built-in helper to check the type of error. `h3` provide [isError](https://unpkg.com/browse/h3@0.7.8/dist/index.d.ts#L126) but it check `instanceof` and does't work with response on client-side. 

```ts
// Mandatory imports
import { FetchError } from 'ohmyfetch'
import type { H3Error } from 'h3'

try { $fetch() } catch (e) {

  // Type-safe check error type
  // You need implement your own
  const isH3FetchError = (e: any): e is FetchError<H3Error> => {
    return e instanceof FetchError && e.data?.statusMessage !== undefined && e.data?.statusCode !== undefined
  }

}

**Comment by pi0** [comment]
Hi @cawa-93, and thanks for the suggestions. 

Regarding return type, you can always cast it and also handle errors. Error is not returned but it is thrown! If we were sure of all possible string combinations, it could be a nice addition but we are not so it is always the best idea that you catch fetch errors.

Regarding typing body, I'm not sure how we can improve it within Nuxt but you can declare it's type like this:

```js
$fetch('/api/update', { 
 body: <MyRequestBody> {
  id: 'not-a-number' // Should show TS Error
 }
})
```

Regarding H3/Fetch errors, they are completely different. statusMessage and statusCode can originate from any server not essentially H3. Would you please explain more about your use case of `isH3FetchError`?

**Reply by cawa-93** [reply]
Reflecting on this later, I realized that I was only thinking about the case of communication of API components within a single application and did not take into account requests for third-party sources, which obviously can return anything in any format.

So, now I create a separate file with interfaces for each api point:
```ts
// server/api/type-api.contracts.ts
export interface TypeApiGetRequest extends FetchOptions {
  method: 'GET'
  params: {
    id: number
  }
}

export interface TypeApiPostRequest extends FetchOptions {
  method: 'POST'
  body: {
    id: number
  }
}

export type TypedRequest = TypeApiGetRequest | TypeApiPostRequest
```

In vue component I import TypedRequest type and use it before $fetch:
```vue
<script setup lang="ts">
const opt: TypedRequest = {
  method: 'POST',
  body: {
    id: 123 // type-safe !
  },
}
const a = await $fetch('/api/type-api', opt)
</script>
```
Finally, on the server, in the request handler, I read the body
```ts
// server/api/type-api.ts
export default defineEventHandler(async (event) => {
  const body = await useBody<TypeApiPostRequest['body']>(event)
  return body.id + 1 // type-safe !
})
```
But unlike the `useBody`, `useQuery` is not generic
```ts
const params = useQuery<TypeApiPostRequest['params']>(event) // Will't work 
```

Regarding errors. 
All my api endpoints use `sendError` with `createError` from `h3` for un-success responses. 
```ts
  if (isNaN(parseInt(id, 10))) {
    return sendError(event,
      createError({
        statusCode: constants.HTTP_STATUS_BAD_REQUEST,
        statusMessage: 'ID is not number',
      }),
    )
  }
```
It seems pretty standardized.
What I wanted was a type-safe process to handle unsuccessful responses from the server in its components.
```ts
try {
$fetch('/api/id', {body: { id: 'nan' }})
}
catch(error: unknown) {
// The default error is of type unknown. Therefore, it is necessary to bring it to a known type.
// $fetch Automatically throw errors when response.ok is false with a friendly error message and compact stack (hiding internals).
// Parsed error body is available with error.data. You may also use FetchError type.

if (error instanceof FetchError) {
  // Now error has FetchError type. But error.data still unknown.
  // h3 always intercepts all errors and brings them to the structure of H3Error

  // I could use isError from h3. However, the client loses information about the instance.
  isError(error.data) // false 

  // Therefore, as an alternative, I simply check for two required fields. (https://github.com/unjs/h3/blob/417af22d737832133f0d7025e29d915b11ff8881/src/error.ts#L15-L16)

  if (error.data?.statusMessage !== undefined && error.data?.statusCode !== undefined) {
    // Now error has FetchError type with the body in error.data most likely has type H3Error type
    showErrorMessageInComponent(
      error.data.statusMessage // type-safe
    )
  }
}

}
```

**Comment by cawa-93** [comment]
Looks like [trpc-nuxt](https://github.com/wobsoriano/trpc-nuxt) provides exactly what I looking for: End-to-end typesafe APIs 🤯
